Support for load balancers specified by hostname

alangibson01 opened the pull request:
Picking up where Shin left off with this work in August (see
https://github.com/sky-uk/feed/pull/153 for how far he got, but
I thought it was best to do this as a fresh PR, since quite some
time has passed).

Edited highlights are that it's now possible to specify hostnames
for internal and external load balancers, and have Feed DNS set up
CNAMEs to those for ingresses, rather than being restricted to
working with AWS load balancers.

I've done some fairly substantial refactoring, to separate the
common parts of the processing loops from those which are specific
to creating CNAMEs to load balancer hostnames and aliases to AWS
load balancers.

sebbonnet commented on controller/controller.go:
```diff
@@ -22,6 +22,7 @@ import (
 
 const ingressAllowAnnotation = "sky.uk/allow"
+const frontendSchemeAnnotation = "sky.uk/frontend-scheme"
 const frontendElbSchemeAnnotation = "sky.uk/frontend-elb-scheme"
```
should we mark this as deprecated?

alangibson01 replied:
Yes we should, I forgot to do that.

sebbonnet commented on dns/dns_updater.go:
```diff
@@ -167,18 +103,39 @@ func (u *updater) Update(entries controller.IngressEntries) error {
 	var nonManaged []string
 	for _, rec := range rrs {
-		if rec.AliasTarget != nil && rec.AliasTarget.DNSName != nil && managedLBs[*rec.AliasTarget.DNSName] {
+		if rec.name != "" && managedLBs[rec.pointsTo] {
```
records with no dns name used to be excluded `rec.AliasTarget.DNSName != nil`, but I saw no test for that, so I'm unsure whether this is still needed

alangibson01 replied:
CNAME records won't have an alias target, so that check isn't appropriate any more.

sebbonnet replied:
this check was done on A records.
However looking at the ResourceRecordSet.AliasTarget doc, I can see that DNSName is a required field, so that check was redundant

sebbonnet commented on dns/dns_updater.go:
```diff
@@ -189,7 +146,7 @@ func (u *updater) determineManagedRecordSets(rrs []*route53.ResourceRecordSet) [
 }
 
-func (u *updater) calculateChanges(originalRecords []*route53.ResourceRecordSet,
+func (u *updater) calculateChanges(originalRecords []consolidatedRecord,
```
We should also update the CNAME when its TTL has changed

alangibson01 commented:
@sebbonnet I've made the change you suggested, please take another look.

alangibson01 commented on cmd/feed-dns/main.go:
```diff
@@ -81,7 +91,18 @@ func main() {
 
-	dnsUpdater := dns.New(r53HostedZone, elbRegion, elbLabelValue, albNames, awsAPIRetries)
+	var dnsUpdater controller.Updater
+	if internalHostname != "" && externalHostname != "" {
```
Indeed we can.

alangibson01 commented:
@sebbonnet I've just responded to that, and made the other changes you recommended, but which I forgot to do.

alangibson01 commented:
@sebbonnet @rewiko I've made a few more changes to clean stuff up, and add an example deployment. Can one or both of you have another look, please?

jsravn commented on dns/static_hostname_adapter.go:
```diff
@@ -0,0 +1,52 @@
+}
+
+func (s staticHostnameAdapter) newChange(action string, host string, details dnsDetails) *route53.Change {
+	set := &route53.ResourceRecordSet{
```
I'd probably call this `rrs` for resource records.

alangibson01 replied:
Done.

alangibson01 commented on cmd/feed-dns/main.go:
```diff
@@ -67,6 +71,12 @@ func init() {
+	flag.StringVar(&externalHostname, "external-hostname", "",
+		"hostname of the internet facing load-balancer. If specified, internal-hostname must also be given.")
+	flag.DurationVar(&cnameTimeToLive, "cname-ttl", defaultCnameTTL,
+		"time-to-live of CNAME records")
```
Done

alangibson01 commented on cmd/feed-dns/main.go:
```diff
@@ -99,9 +113,28 @@ func main() {
 	select {}
 }
 
+func createLoadBalancerAdapter() (dns.LoadBalancerAdapter, error) {
```
Done.

alangibson01 commented on cmd/feed-dns/main.go:
```diff
@@ -99,9 +113,28 @@ func main() {
+	if (internalHostname == "" && externalHostname != "") || (internalHostname != "" && externalHostname == "") {
+		log.Error("Must supply both internal-hostname and external-hostname if any are to be provided.")
+		os.Exit(-1)
+	}
```
This was a carry-over from the original PR, and I think you raised exactly the same question. Shin's response at that time was:

> It still applies if you have other internal services that need a more direct route and your external needs to be fronted by a CDN for example.
> We could however make them optional i.e. have one or the other (or both), would that make more sense?

What are your thoughts on that?

jsravn replied:
I think making them optional makes sense, at least in m25 we'll only have one frontend for now.

alangibson01 commented on dns/aws_adapter.go:
```diff
@@ -0,0 +1,132 @@
+	}, nil
+}
+
+func (a awsAdapter) initialise(schemeToDNS map[string]dnsDetails) error {
```
You're right, I guess that's what years of working with pass-by-reference languages does to you.

alangibson01 commented on dns/dns_updater.go:
```diff
@@ -6,52 +6,46 @@ import (
-	DescribeLoadBalancers(input *aws_alb.DescribeLoadBalancersInput) (*aws_alb.DescribeLoadBalancersOutput, error)
+// LoadBalancerAdapter defines operations which vary based on the type of load balancer being used for ingress.
+type LoadBalancerAdapter interface {
+	initialise(schemeToDNS map[string]dnsDetails) error
```
Yes and yes.

alangibson01 commented on dns/dns_updater.go:
```diff
@@ -6,52 +6,46 @@ import (
+	initialise(schemeToDNS map[string]dnsDetails) error
+	newChange(action string, host string, details dnsDetails) *route53.Change
+	changeExistingIfRequired(record consolidatedRecord, host string, details dnsDetails) *route53.Change
 }
```
I think it probably does make sense to have a separate package for the adapters.

alangibson01 replied:
Makes things a bit messy for the tests, but maybe that's a sign the tests aren't right.

alangibson01 commented on dns/dns_updater.go:
```diff
@@ -259,36 +222,18 @@ func (u *updater) createChanges(hostToIngress hostToIngress,
+		if existingRecord, recordExists := indexedRecords[recordKey{host, dnsDetails.dnsName}]; !recordExists {
+			changes = append(changes, u.lbAdapter.newChange("UPSERT", host, dnsDetails))
+		} else if changedRecord := u.lbAdapter.changeExistingIfRequired(existingRecord, host, dnsDetails); changedRecord != nil {
+			changes = append(changes, changedRecord)
```
That's a good idea.

alangibson01 commented on dns/dns_updater.go:
```diff
@@ -241,14 +204,14 @@ func (u *updater) indexByHost(entries []controller.IngressEntry) (hostToIngress,
+	originalRecords []consolidatedRecord) ([]*route53.Change, []string) {
 
 	type recordKey struct{ host, elbDNSName string }
 	var skipped []string
```
Done

alangibson01 commented on dns/dns_updater.go:
```diff
@@ -167,18 +105,43 @@ func (u *updater) Update(entries controller.IngressEntries) error {
+				record.ttl = *recordSet.TTL
+			}
+			records = append(records, record)
+		}
```
Yes, we can give each of the adapters a chance to recognise a record, and create a "consolidated" record from it.

alangibson01 commented:
@jsravn The more I think about it, the more moving the adapters into a package of their own makes sense. One of the benefits will be that it enforces a clearer separation between the logic of the updater, and the nuts and bolts of actually interacting with ELBs and creating the right kinds of record. As a knock-on effect, we should be able to simplify some of the testing and avoid having so much repetition.

alangibson01 commented on cmd/feed-dns/main.go:
```diff
@@ -99,9 +113,28 @@ func main() {
+	if (internalHostname != "" || externalHostname != "") && elbLabelValue != "" {
+		log.Error("Can't supply both ELB label and non-ELB hostname. Choose one or the other.")
+		os.Exit(-1)
+	}
```
Good point.

alangibson01 commented:
@jsravn My latest push addresses all of your comments, except the one about whether we need the internal/external facility for a non-AWS frontend, and the one about moving the adapters into their own package.

As I said above, in theory I agree with pushing the adapters into their own package, since it would keep the interface between the update control loop and the specific frontends nice and clean. However, because the tests in `dns_updater_test.go` need to overwrite some of the fields of `awsAdapter` in order to inject mocks, either I need to do a bit of an overhaul of the tests to avoid that, or enable it by breaking the encapsulation we introduce with the move of the adapters into their own package.

I'm happy to do the work overhauling the tests - anything to do with the interaction with the specific frontend would go in new test files in the adapter package, with the existing tests being changed to use a dedicated stub adapter that it more test-friendly. I think this could provide benefits in terms of reducing duplication and making the existing tests more readable, but it is rather a large set of changes. What do you think?

alangibson01 commented:
Actually, forget my previous comments. I think there are better ways for me to move the adapters into their own package without having to change the current tests, which would make them more unit-y.

alangibson01 commented:
@jsravn Please take another look now, I've finished the refactoring and make internal and external names optional, as requested.

jsravn commented on dns/adapter/frontend_adapter.go:
```diff
@@ -0,0 +1,24 @@
+type FrontendAdapter interface {
+	Initialise() (map[string]DNSDetails, error)
+	CreateChange(action string, host string, details DNSDetails, recordExists bool, existingRecord *ConsolidatedRecord) *route53.Change
+	Recognise(*route53.ResourceRecordSet) (*ConsolidatedRecord, bool)
```
Not quite sure what this does from the method name. Maybe add a comment?

jsravn replied:
Maybe rename to something like 'IsManaged'.
